Add more shell script lints

aneeshusa opened the pull request:
<!-- Please describe your changes on the following line: -->

The "$(some_command arg1 arg2)" form is preferred to the
`some_command arg1 arg2` form because it nests unambiguously.
Add a lint for this to tidy.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12736)

<!-- Reviewable:end -->

aneeshusa committed:
Add lint for backticks in shell scripts

The "$(some_command arg1 arg2)" form is preferred to the
`some_command arg1 arg2` form because it nests unambiguously.
Add a lint for this to tidy.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/tidy/servo_tidy_tests/test_tidy.py, python/tidy/servo_tidy/tidy.py, python/tidy/servo_tidy_tests/shell_tidy.sh

aneeshusa committed:
Add lint to ensure substitutions use the full form

Check that any variable substitutions use the full ${VAR} form,
not just $VAR (but don't check for quoting yet).

emilio commented:
Seems like this is going to need to fix the offending scripts before landing:

```
./components/servo/fake-ld.sh:12: variable substitutions should use the full "${VAR}" form
./etc/ci/lockfile_changed.sh:12: variable substitutions should use the full "${VAR}" form
./etc/ci/lockfile_changed.sh:13: variable substitutions should use the full "${VAR}" form
./etc/ci/manifest_changed.sh:20: variable substitutions should use the full "${VAR}" form
./etc/ci/manifest_changed.sh:21: variable substitutions should use the full "${VAR}" form
./etc/ci/upload_docs.sh:15: variable substitutions should use the full "${VAR}" form
./etc/ci/upload_nightly.sh:30: variable substitutions should use the full "${VAR}" form
./etc/ci/upload_nightly.sh:61: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/regen.sh:11: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/regen.sh:12: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/regen.sh:32: script should not use backticks for command substitution
./ports/geckolib/gecko_bindings/tools/regen.sh:35: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/setup_bindgen.sh:12: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/setup_bindgen.sh:18: script should not use backticks for command substitution
./ports/geckolib/gecko_bindings/tools/setup_bindgen.sh:28: variable substitutions should use the full "${VAR}" form
./ports/geckolib/gecko_bindings/tools/setup_bindgen.sh:29: variable substitutions should use the full "${VAR}" form
```

aneeshusa committed:
Add lint for `[` instead of `[[` in shells scripts

aneeshusa commented:
Updated with another lint. For some reason `./mach test-tidy` didn't catch those errors, I had to run `./mach test-tidy --all` for it to find them.

Wafflespeanut commented:
Sorry I didn't look into this earlier. FYI, `./mach test-tidy` only checks the files in the diff. :)
The changes LGTM. @bors-servo r+

bors-servo commented:
:pushpin: Commit f07d8f1 has been approved by `Wafflespeanut`

<!-- @bors-servo r=Wafflespeanut f07d8f188a0621545957a0f348271d29b08704ec -->

bors-servo commented:
:hourglass: Testing commit f07d8f1 with merge aa900b9...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2315), [arm64](http://build.servo.org/builders/arm64/builds/2324), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2371), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2499), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2403), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2386), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2461), [windows-dev](http://build.servo.org/builders/windows-dev/builds/269)
